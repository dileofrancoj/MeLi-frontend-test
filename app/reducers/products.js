import {
  FETCH_ERROR,
  FETCH_SUCCESS,
  FETCHING,
} from "./../reducers/actions/common";
export const initialState = {
  loading: false,
  error: false,
  products: [],
  categories: [],
};

export function productsReducer(state, action) {
  switch (action.type) {
    case FETCHING:
      return {
        loading: true,
        error: false,
        product: [],
        categorie: [],
      };
    case FETCH_SUCCESS:
      return {
        loading: false,
        error: false,
        products: action.payload.items,
        categories: action.payload.categories,
      };
    case FETCH_ERROR:
      return {
        loading: false,
        error: true,
        products: [],
        categories: [],
      };
    default:
      return {
        state,
      };
  }
}
