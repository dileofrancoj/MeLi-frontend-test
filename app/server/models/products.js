import fetch from "isomorphic-fetch";
import { BASE_URL } from "../utils/constants";

const list = async (q, limit = 4) => {
  const data = await fetch(
    `${BASE_URL}/sites/MLA/search?q=${q}&limit=${limit}`
  );
  return await data.json();
};

const single = async (id) => {
  const data = await fetch(`${BASE_URL}/items/${id}`);
  return await data.json();
};

const singleDescription = async (id) => {
  const data = await fetch(`${BASE_URL}/items/${id}/description`);
  return await data.json();
};

module.exports = { list, single, singleDescription };
