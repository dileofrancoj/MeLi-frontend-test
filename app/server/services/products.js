const sendProducts = (_products) =>
  _products.map(
    ({
      id,
      title,
      currency_id: currency,
      price,
      thumbnail: picture,
      condition,
      shipping,
      address,
    }) => {
      return {
        id,
        title,
        price: {
          currency,
          amount: price,
          decimals: 2,
        },
        picture,
        condition,
        free_shipping: shipping.free_shipping,
        address: address.state_name,
      };
    }
  );

const sendProduct = ({ product: _product, description, categories }) => {
  return {
    id: _product.id,
    title: _product.title,
    price: {
      currency: _product.currency_id,
      amount: _product.price,
      decimals: 2,
    },
    picture: _product.pictures[0].secure_url,
    condition: _product.condition,
    free_shipping: _product.shipping.free_shipping,
    sold_quantity: _product.sold_quantity,
    description,
    categories,
  };
};

module.exports = { sendProducts, sendProduct };
