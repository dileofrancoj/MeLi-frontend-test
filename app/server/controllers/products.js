const { list, single, singleDescription } = require("../models/products");
import { single as singleCat } from "./../models/categories";
const {
  filterToCategories,
  categoriesSingle,
} = require("../services/categories");
const { sendProducts, sendProduct } = require("../services/products");

const author = {
  name: "Franco",
  lastname: "Di Leo",
};

const searchById = async (id) => {
  try {
    const obj = {};
    const product = await single(id);
    const { category_id } = product;
    const { path_from_root } = await singleCat(category_id);

    const categories = filterToCategories(path_from_root);

    const { plain_text: description } = await singleDescription(id);
    // ConfiguraciÃ³n de Babel no me permite usar spread
    Object.assign(obj, { product, description, categories });
    const item = sendProduct(obj);
    return item;
  } catch (e) {
    console.error(e);
  }
};

const search = async (product) => {
  try {
    const { results, filters } = await list(product);
    //if (results.length === 0) return res.sendStatus(404);
    const items = sendProducts(results);
    const categories = filterToCategories(filters);
    return { author, items, categories };
    //res.json({ author, categories, items });
  } catch (e) {
    console.error(e);
  }
};

module.exports = { search, searchById };
