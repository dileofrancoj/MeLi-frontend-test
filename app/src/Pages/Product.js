import React, { useEffect, useReducer } from "react";
import { Row, Col, Button } from "react-bootstrap";
import { searchById } from "../../server/controllers/products";
import { formatAsCurrency } from "../../server/utils/helpers";
import { useParams } from "react-router-dom";

/* Personal components */
import Loading from "../Components/Loading";
import Breadcrumb from "../Components/Breadcrumb";

import { productReducer, initialState } from "../../reducers/product";
import { FETCH_SUCCESS, FETCH_ERROR } from "../../reducers/actions/common";
const Product = () => {
  const { id } = useParams();

  const [{ product, loading }, dispatch] = useReducer(
    productReducer,
    initialState
  );
  const getProduct = async (id) => {
    try {
      const data = await searchById(id);
      dispatch({ type: FETCH_SUCCESS, payload: data });
      //setProduct(data);
    } catch (e) {
      console.error(e);
      dispatch({ type: FETCH_ERROR, payload: {} });
    }
  };
  useEffect(() => {
    getProduct(id);
  }, []);

  const RenderProduct = () => {
    return (
      <React.Fragment>
        <Breadcrumb categories={product.categories} />
        <section className="vh-75 bg-white ">
          <Row className="justify-content-center pt-32 bg-white m-0">
            <Col md={8} className="item-header text-center">
              <picture className="mt-32">
                {/* Formatos */}
                <img src={product.picture} alt={product.title} />
              </picture>
            </Col>
            <Col md={4} className="pr-32">
              <p className="f-14">
                {product.condition === "new" ? "Nuevo" : "Usado"} -{" "}
                {product.sold_quantity} Vendidos
              </p>
              <h4 className="mt-16 f-24">{product.title}</h4>
              <h3 className="mt-32 f-46">
                $ {formatAsCurrency(product.price.amount)}
              </h3>
              <Button variant="primary" className="w-100 mt-32 ">
                Comprar
              </Button>
            </Col>
          </Row>
          <Row className="mt-64 pl-32">
            <Col className="pl-32" md={8}>
              <h4 className="f-28 text-left ">Descripci√≥n del producto</h4>
              <p className="mt-32 mb-32 ">{product.description}</p>
            </Col>
          </Row>
        </section>
      </React.Fragment>
    );
  };

  return (
    <React.Fragment>
      {/* Se muestra si product = true && loading = false */}
      {!!product && !loading ? <RenderProduct /> : <Loading />}
    </React.Fragment>
  );
};

export default Product;
