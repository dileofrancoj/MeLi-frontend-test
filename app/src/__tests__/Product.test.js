import React from "react";
import { shallow } from "enzyme";
import { StaticRouter } from "react-router-dom";
import { searchById } from "../../server/controllers/products";

import Product from "./../Pages/Product";

describe("Product", () => {
  test("should render basic Product component ", () => {
    const component = shallow(
      <StaticRouter>
        <Product />
      </StaticRouter>
    );
    expect(component).toMatchSnapshot();
  });
  test("should get an Object by Id", async () => {
    const productId = "MLA867086863";
    const data = await searchById(productId);
    expect(typeof data).toBe("object");
  });
});
