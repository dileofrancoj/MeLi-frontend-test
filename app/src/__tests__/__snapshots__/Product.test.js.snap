// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render basic Product component  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <Product />
</Router>
`;
